### ISSUE TITLE
Fetching large size of messages is quite slow

### ISSUE THREAD

**Issue by jiaminghong** [tulios/kafkajs#841]
Hi, I'm working on a project recently with kafkajs, but encountered some obstacles. Could you can give me some hands.

### Reproduce
1. Read messages from one topic with 16 partitions.
2. Filter out unqualified messages.
3. Bulk write into database.

### Some backgrounds about the project
1. The volume of message lying in each partition is quite large, around 500 million across all the partitions and the size of each message is around 2000 bytes on average.
2. Bulk write into database will take quite long time, about 50 seconds for 15k messages.
3. I can only read from one partition before the write-into-database process is resolved, otherwise the written operation will stomp on each other, so ``partitionsConsumedConcurrently = 1`` is a must.

### Possible solutions
1. The first hint came to my mind is that I can set ``minBytes``, ``maxBytes``, ``maxBytesPerPartition`` and ``sessionTimeout`` as large as possible to fetch more data from kafka which results in less database written operation: ``minBytes=30,000,000  maxBytes=50,000,000  maxBytesPerPartition=50,000,000  sessionTimeout=180,000``, but the problem is fetching 30 million byte messages from partition-x will also take 25 seconds which is not efficient as expected.

2. Set ``minBytes``, ``maxBytes``, ``maxBytesPerPartition`` relatively small number, and use a variable to accumulate the message before writing into database then write into database when the number of messages hits a fixed number like 500,000:
```javascript
const accumulateMessages = [];
await consumer.run({
    partitionsConsumedConcurrently: 1,
    eachBatchAutoResolve: ???
    autoCommit: ???
    eachBatch: async ({ batch, resolveOffset, heartbeat, isRunning, isStale }) => {
        const filteredMessages = filterFunc(batch.messages);
        accumulateMessages.push.apply(accumulateMessages, filteredMessages);
        resolveOffset() ???
        if (accumulateMessages.length === 500000 || consumerWaittingTimeInMs === 1800000) {
            await writeIntoDatabase(accumulateMessages);
            await heartbeat();
            commitOffsetsIfNecessary() ???
            accumulateMessages = [];
            consumerWaittingTimeInMs = 0;
        }
    }
})
```
The problem here is I don't know how to implement the logic to meet the following requirements, as you can see the question marks in the code. **This is the key part that I need help with**.
* We shouldn't commit any offsets unless we can confirm that the message has been insert into database. (we have to make sure every message will be processed)
* But if we never commit any offsets before written operation resolved, will the consumer re-fetch the same messages from a specific partition? (before hitting the fixed number of messages, it might read from same partition several times and start from the same offset submitted previously)
* Whenever the consumer crashes before the written process is done, we shouldn't commit offsets, so we can't use ``resolveOffset()`` under this situation because in case of errors it will commit offsets?

### More questions about kafkajs API
1. How large can ``minBytes``, ``maxBytes``, ``maxBytesPerPartition`` be set to in theory? 10 million, 10 billion?
2. What's the purpose of ``resolveOffset()`` on earth? Is it intended to commit offsets? If not, what's its difference/relation with ``submitOffsets()`` or ``submitOffsetsIfNecessary()``

Really appreciate any helps!!!

**Comment by Nevon** [comment]
> I can only read from one partition before the write-into-database process is resolved, otherwise the written operation will stomp on each other, so partitionsConsumedConcurrently = 1 is a must.

Do you mean that your database does not support concurrent writes? Consuming from a single partition at a time obviously means that you'll lose any parallelism gains. Since most of your time would be spent on IO, it would be ideal to be able to process several partitions concurrently.

> the problem is fetching 30 million byte messages from partition-x will also take 25 seconds which is not efficient as expected.

Have you looked into where you're spending the time? If it's network transfer time, then there's not much that can be done other than moving your application closer to the cluster. We recently merged a pull request that allows you to fetch from specific replicas based on `rackId`. If your consumer is in the same physical location as the broker it's fetching from, this could speed things up.

> Set minBytes, maxBytes, maxBytesPerPartition relatively small number, and use a variable to accumulate the message before writing into database then write into database when the number of messages hits a fixed number like 500,000:

You can definitely do that, although it still sounds strange to me that you can't do multiple concurrent writes instead, so that you can process several partitions in parallel. With these large data volumes, most of your time should be spent just waiting for network, so it sounds ideal to parallelize as long as you have enough bandwidth.

> How large can minBytes, maxBytes, maxBytesPerPartition be set to in theory? 10 million, 10 billion?

In theory it's only limited by the available memory, but that doesn't mean you'll get any performance benefits from it. Node is optimized for asynchronous IO-heavy workloads, not blocking synchronous work, so trying to work with those huge buffers could cause huge event loop lag.

> What's the purpose of resolveOffset() on earth? Is it intended to commit offsets? If not, what's its difference/relation with submitOffsets() or submitOffsetsIfNecessary()

`resolveOffset` is a local, in-memory operation that simply marks an offset as "processed". You can use it to advance the offset for the next fetch. So in your scenario, you could resolve offsets to accumulate messages and only commit once you have successfully written your messages to the database.

`commitOffset` tells the broker that your consumer group has processed an offset, so if your consumer shuts down, it will resume from that offset. `commitOffsetsIfNecessary` does the same thing, except it respects the `autoCommit` settings. So if you are implementing `eachBatch` and want to commit every 100 messages, you can just configure the `autoCommitThreshold` and then call `commitOffsetsIfNecessary` on every message processed, without having to worry about implementing the counter and conditional committing yourself.

> Whenever the consumer crashes before the written process is done, we shouldn't commit offsets, so we can't use resolveOffset() under this situation because in case of errors it will commit offsets?

This is a bug. If `autoCommit` is `false`, we should not commit resolved offsets on errors. In the normal case where you process messages in-order, it makes sense to commit resolved offsets on error so that we don't have to re-process the messages that were already processed, but in your case where you want to commit completely manually, we should not.

**Comment by ankon** [comment]
> * We shouldn't commit any offsets unless we can confirm that the message has been insert into database. (we have to make sure every message will be processed)
> * But if we never commit any offsets before written operation resolved, will the consumer re-fetch the same messages from a specific partition? (before hitting the fixed number of messages, it might read from same partition several times and start from the same offset submitted previously)
> * Whenever the consumer crashes before the written process is done, we shouldn't commit offsets, so we can't use resolveOffset() under this situation because in case of errors it will commit offsets?

I find this problem somewhat interesting to think about, so purely as a "thinking point": Maybe you don't want to commit to Kafka at all, but rather commit into the database (or another storage structure). Kafka wouldn't mind, and you basically end up having to manage your offsets on your own. But it could give you the needed flexibility to know which messages you still need to process at any point in time.

**Comment by jiaminghong** [comment]
@Nevon  Thanks for your quick response! With your nice explanation, I have successfully implemented all the requirements with kafkajs. Before pointing out one more question/concern, I'd love to clarify the question you mentioned.

> Do you mean that your database does not support concurrent writes? Consuming from a single partition at a time obviously means that you'll lose any parallelism gains. Since most of your time would be spent on IO, it would be ideal to be able to process several partitions concurrently. >

### Clarification
My database does support concurrent writes.
However, the actual flow of processing data in my project restricts it: ***fetch data from kafka topic*** --> ***deduplicate the data with same id*** --> ***write data into a temporary table*** --> ***merge into a main table with the temporary table --> clean the temporary table***. I have to make sure data in temporary table is unique otherwise it will definitely fail the merge statement, in other words, consuming concurrently will lead to also insert into database in parallel which might produce duplicate rows in the temporary table.

### Question
With my service running currently with good performance, I noticed that the consumer only consumes data from specific partitions. There are 8 partitions **[0 , 1, 2, 3, 4, 5, 6, 7]** in the topic for my project, the consumer only reads data from, for example, **partition 0**, **6** and **7**, only after the consumer exhausts all the messages from any of **partition 0**, **6** or **7**, it would choose another partition to consume from. 

This behavior does worry me a lot due to the retention time of the topic as well as the large volume of messages, around 200 million, in the each partition. What if the message in the beginning of **partition 2** hits the retention time,  and the consumer is still "diligently" consuming from other partitions, finally it's **partition 2**'s turn to be consumed from, but it turns out the latest commit offset of **partition 2** doesn't exist any more, which throws out the error: 
``"error":"The requested offset is not within the range of offsets maintained by the server"``

**Is there any option in kafkajs or any other way that I can force the consumer to fetch data from every partition rather than some of them?**

**Comment by Nevon** [comment]
> fetch data from kafka topic --> deduplicate the data with same id --> write data into a temporary table --> merge into a main table with the temporary table --> clean the temporary table. I have to make sure data in temporary table is unique otherwise it will definitely fail the merge statement, in other words, consuming concurrently will lead to also insert into database in parallel which might produce duplicate rows in the temporary table.

Now, I don't know the exact circumstances for your problem, but just to provide you with some information that may simplify your life:

1. If the events are keyed by the aggregate id (for example, if the events are for a user, the key could be the user id), all events will end up on the same partition. This means that all events for that user will be in order on the same partition and are guaranteed to be consumed in order. If this is the case, then even if you have 16 partitions you could consume all 16 partitions in parallel without risking the events being out of order.
2. If your topic contains "versions" of a certain aggregate, such as updated versions of an order, you can use the order id as the key on a compacted topic, which would mean that Kafka itself could do the merging and there would only ever be one message for a given key. I.e. Kafka could do the deduplication instead of having to do this aggregation in your application.

> There are 8 partitions [0 , 1, 2, 3, 4, 5, 6, 7] in the topic for my project, the consumer only reads data from, for example, partition 0, 6 and 7, only after the consumer exhausts all the messages from any of partition 0, 6 or 7, it would choose another partition to consume from.

This is decided by the broker, and I'm actually not sure how this logic works. From the client side, we make requests for all assigned partitions, but it's up to the brokers to decide how they decide to allocate the requested number of bytes across the different partitions. There may be some way to configure this on the broker side, but I have not come across it.

At the end of the day, the right way to handle it is to make sure that you can horizontally scale across the partitions in the topic. Designing your application to have a single consumer consume from all partitions is a losing battle. What happens when the incoming number of messages per second is higher than you can process in a second?

**Comment by jiaminghong** [comment]
@Nevon That's great! your solution perfectly solved my question. Really appreciate your help since I have been upset for several days due to not seeking out any useful solutions on how to fairly allocate partitions.

**Comment by Nevon** [comment]
Glad that I could help. I'm gonna close this issue now in favor of #849
